prov/verbs/src/verbs_domain.c

```diff
@@ -56,12 +80,28 @@ static void vrb_set_credit_handler(struct fid_domain *domain_fid,
 	domain->send_credits = credit_handler;
 }
 
-static int vrb_enable_ep_flow_ctrl(struct fid_ep *ep_fid)
+static bool vrb_flow_ctrl_available(struct fid_ep *ep_fid)
 {
 	struct vrb_ep *ep = container_of(ep_fid, struct vrb_ep, util_ep.ep_fid);
+
 	// only enable if we are not using SRQ
-	if (!ep->srq_ep && ep->ibv_qp && ep->ibv_qp->qp_type == IBV_QPT_RC) {
-		ep->peer_rq_credits = 1;
+	return (!ep->srq_ep && ep->ibv_qp && ep->ibv_qp->qp_type == IBV_QPT_RC);
+}
+
+static int vrb_enable_ep_flow_ctrl(struct fid_ep *ep_fid)
+{
+	struct vrb_ep *ep = container_of(ep_fid, struct vrb_ep, util_ep.ep_fid);
+	struct vrb_cq *cq = container_of(ep->util_ep.rx_cq, struct vrb_cq, util_cq);
+
+	if (vrb_flow_ctrl_available(ep_fid)) {
+		ofi_genlock_lock(&cq->util_cq.cq_lock);
+		/*
+		 * peer_rq_credits is initialized to UINT64_MAX (same as -1).
+		 * It is possible credit updates arrived earlier. The new
+		 * credit should be (received credits + 1).
+		 */
+		ep->peer_rq_credits += 2;
+		ofi_genlock_unlock(&cq->util_cq.cq_lock);
```

Comment:
I find this confusing.  The credit scheme overall isn't great, at it uses a counter based mechanism (+10 new messages can now be sent), rather than a sliding window sort of scheme (send up to message ID 400).  This results in weird initialization like rq_credits_avail = -1, and the non-intuitive setting of peer_rq_credits += 2 when enabling.

I think there's a good chance this would be simpler if we switched to a window credit scheme instead.  Then we could have something like:
```
if (peer_rq_credits == UINT64_MAX)
    peer_rq_credits = 1;
```
Even if we kept to a counter scheme, I would have expected the above flow to be:
```
if (peer_rq_credits == UINT64_MAX)
    peer_rq_credits = 1;
else
    peer_rq_credits++;  /* both sides assume 1 credit to start with */
```
I understand why the +2 is in the above code, but it's not intuitive at all and seems likely to confuse someone else down the line.

Bad news is that changing the credit scheme changes the protocol.

Reply:
Actually, this changes the CM protocol anyway.  We should handle that.  Gah.

Reply:
One way is to use a different variable to track the received credit before flow control is enabled:
```
peer_rq_credits = saved_peer_credit + 1;  /* both sides assume 1 credit to start with */
```

Reply:
That sounds good.  I think we want to keep protocol compatibility.  That requires changing the value of flow_ctrl mentioned above.